include/AudioPortsSettings.h

```diff
@@ -0,0 +1,162 @@
+
+} // namespace detail
+
+
+//! Metafunction to convert `AudioDataKind` to its type
+template<AudioDataKind kind>
+using GetAudioDataType = typename detail::GetAudioDataTypeHelper<kind>::type;
+
+
+//! Compile-time customizations for audio ports
+struct AudioPortsSettings
+{
+	//! The audio data type used by the processor
+	AudioDataKind kind;
+
+	//! The audio data layout: interleaved or non-interleaved
+	bool interleaved;
+
+	//! The number of input channels, or `DynamicChannelCount` if unknown at compile time
+	proc_ch_t inputs = DynamicChannelCount;
+
+	//! The number of output channels, or `DynamicChannelCount` if unknown at compile time
+	proc_ch_t outputs = DynamicChannelCount;
+
+	/**
+	 * In-place processing
+	 *
+	 * When true, the processor always uses in-place processing and its process method
+	 *   will only need a single in/out buffer parameter rather than two separate parameters.
+	 *
+	 * When false, the processor may be dynamically in-place if the audio buffer implementation
+	 *   supports it, otherwise the processor never uses in-place processing.
+	 *
+	 *   In order to support dynamically in-place processing, an audio buffer implementation can
+	 *   return the same buffer for both its input buffer and its output buffer. Any processors
+	 *   using such a buffer implementation must design their process method to check whether the
+	 *   input and output buffers are the same buffer (indicating in-place processing is active),
+	 *   and safely handle both possibilities.
+	 *
+	 * NOTE: Regardless of whether `inplace` is true or false, if `buffered` is true, the process method
+	 *   will have no buffer parameters.
+	 *
+	 * TODO: Once `RemotePluginAudioPorts` supports statically in-place processing, replace this setting with
+	 *       `forceInplace` and use statically in-place processing when `forceInplace || inputs == 0 || outputs == 0`.
+	 */
+	bool inplace = false;
+
+	/**
+	 * Audio buffer usage
+	 *
+	 * When true, the processor's audio buffers are always written to and read from when routing
+	 *   to and from the processor. This lessens the extent to which the audio port can apply the
+	 *   "direct routing" optimization, but some processors with custom audio buffers (such as Vestige)
+	 *   have no choice because they require their buffers to always be written to and read from.
+	 *
+	 *   For processors that manage their own buffers, passing those buffers to itself through
+	 *   the process method is pointless, so this option also removes the buffer parameter(s)
+	 *   from the process method. TODO: Remove this feature/quirk?
+	 *
+	 * When false, the processor's audio buffers may not always be written to or read from when routing
+	 *   to and from the processor. This can allow for better performance from the "direct routing"
+	 *   optimization, though it may not be suitable for all processor implementations.
+	 */
+	bool buffered = true;
```

Comment:
When implementing this for Lv2, I was wondering how to actually find the correct signature for `processImpl` (after I have specified my own `AudioPortsSettings`). There are a couple of ways:

1. Look at another plugin (not really professional)
2. Write the function and try to infer it from compiler errors (better, but still weird, for example, the compiler told me there was a mismatch and I should use `PlanarBufferView<..., some very high integer>`), which of course corresponds to `DynamicChannelCount`)
3. Trying to infer it from static code analysis :sweat_smile: 

All in all, I guess it would be better to reflect which combinations of which settings cause exactly which signatures.

Reply:
I simplified the `processImpl` signature further in https://github.com/LMMS/lmms/pull/7459/commits/2356774e480613bc8f4dd59192c65cde1a367601.

Then I added some documentation and an extra validation check in https://github.com/LMMS/lmms/pull/7459/commits/4f8b05b95f97b49cf3024d93ba872b2c99672914 and https://github.com/LMMS/lmms/pull/7459/commits/87acc22c0887e3d725cb57331c3ab32d79559f84.